src/apple/deps/mod.rs

```diff
@@ -33,36 +33,159 @@ pub enum Error {
+            bossy::Command::impure_parse("sudo gem install")
+                .with_arg(package)
+                .run_and_wait()
+                .map_err(|source| Error::InstallFailed { package, source })?;
+        }
+        Ok(())
+    }
 }
 
-pub fn install(package: &'static str, reinstall_deps: opts::ReinstallDeps) -> Result<bool, Error> {
-    let found = util::command_present(package)
-        .map_err(|source| Error::PresenceCheckFailed { package, source })?;
-    log::info!("package `{}` present: {}", package, found);
-    if !found || reinstall_deps.yes() {
-        println!("Installing `{}`...", package);
-        // reinstall works even if it's not installed yet, and will upgrade
-        // if it's already installed!
-        bossy::Command::impure_parse("brew reinstall")
-            .with_arg(package)
-            .run_and_wait()
-            .map_err(|source| Error::InstallFailed { package, source })?;
-        Ok(true)
+fn update_package(package: &'static str, gem_cache: &mut GemCache) -> Result<(), Error> {
+    if installed_with_brew(package) {
+        brew_reinstall(package)?;
     } else {
-        Ok(false)
+        gem_cache.reinstall(package)?;
     }
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum PackageSource {
+    Brew,
+    BrewOrGem,
+}
+
+#[derive(Debug)]
+pub struct PackageSpec {
+    pub pkg_name: &'static str,
+    pub bin_name: &'static str,
+    pub package_source: PackageSource,
 }
 
+impl PackageSpec {
+    pub const fn brew(pkg_name: &'static str) -> Self {
+        Self {
+            pkg_name,
+            bin_name: pkg_name,
+            package_source: PackageSource::Brew,
+        }
+    }
+
+    pub const fn brew_or_gem(pkg_name: &'static str) -> Self {
+        Self {
+            pkg_name,
+            bin_name: pkg_name,
+            package_source: PackageSource::BrewOrGem,
+        }
+    }
+
+    pub const fn with_bin_name(mut self, bin_name: &'static str) -> Self {
+        self.bin_name = bin_name;
+        self
+    }
+
+    pub fn found(&self) -> Result<bool, Error> {
+        let found =
+            util::command_present(self.bin_name).map_err(|source| Error::PresenceCheckFailed {
+                package: self.pkg_name,
+                source,
+            })?;
+        log::info!("package `{}` present: {}", self.pkg_name, found);
+        Ok(found)
+    }
+
+    pub fn install(
+        &self,
+        reinstall_deps: opts::ReinstallDeps,
+        gem_cache: &mut GemCache,
+    ) -> Result<bool, Error> {
+        if !self.found()? || reinstall_deps.yes() {
+            println!("Installing `{}`...", self.pkg_name);
+            match self.package_source {
+                PackageSource::Brew => brew_reinstall(self.pkg_name)?,
+                PackageSource::BrewOrGem => update_package(self.pkg_name, gem_cache)?,
+            }
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+}
+
+static PACKAGES: &[PackageSpec] = &[
```

Comment:
I make an exception for `static` items, which I like to have at the very top (below uses and mods) to make globals seem appropriately global